File: examples/todo_app/zig_todo/todo_app.zig
//! Todo app: Vue UI + SQLite + nativeview (Zig port of nim_todo/todo_app.nim).
//! SPDX-License-Identifier: Apache-2.0

const std = @import("std");
const nv = @import("nv");
const ts = @import("ts");
const br = @import("br");
const embed = @import("embed");

const AppCtx = struct {
    store: *ts.TodoStore,
};

var g_ctx: AppCtx = undefined;

fn emitToJs(userdata: ?*anyopaque, event: []const u8, json_body: []const u8) void {
    const win: ?*nv.nv_window_t = @ptrCast(userdata);
    const evz = std.heap.c_allocator.dupeSentinel(u8, event, 0) catch return;
    defer std.heap.c_allocator.free(evz);
    const jz = std.heap.c_allocator.dupeSentinel(u8, json_body, 0) catch return;
    defer std.heap.c_allocator.free(jz);
    nv.nv_send(win, evz.ptr, jz.ptr);
}

fn onReady(window: ?*nv.nv_window_t, userdata: ?*anyopaque) callconv(.c) void {
    _ = window;
    _ = userdata;
    std.debug.print("[todo_app] webview ready\n", .{});
}

fn onMessage(window: ?*nv.nv_window_t, event: ?[*:0]const u8, json_msg: ?[*:0]const u8, userdata: ?*anyopaque) callconv(.c) void {
    const ctx = userdata orelse return;
    const c: *AppCtx = @ptrCast(@alignCast(ctx));
    if (json_msg == null) return;
    const ev = event orelse return;
    if (!std.mem.eql(u8, std.mem.span(ev), "todo")) return;
    const body = std.mem.span(json_msg.?);
    const allocator = std.heap.page_allocator;
    _ = br.bridgeHandleWire(c.store, allocator, body, window, emitToJs);
}

fn onClose(win: ?*nv.nv_window_t, userdata: ?*anyopaque) callconv(.c) void {
    _ = win;
    const app: ?*nv.nv_app_t = @ptrCast(@alignCast(userdata));
    nv.nv_app_quit(app);
}

pub fn main(minimal: std.process.Init.Minimal) void {
    var arg_it = std.process.Args.Iterator.init(minimal.args);
    defer arg_it.deinit();
    _ = arg_it.skip();
    const default_db: [:0]const u8 = "./todo_app.db";
    const db_path: [:0]const u8 = arg_it.next() orelse default_db;

    const store = ts.storeOpen(db_path) orelse {
        std.debug.print("[todo_app] todo_store_open failed for {s}\n", .{db_path});
        return;
    };
    defer ts.storeClose(store);

    const app = nv.nv_app_create() orelse {
        std.debug.print("[todo_app] nv_app_create failed\n", .{});
        return;
    };
    defer nv.nv_app_destroy(app);

    const cfg = nv.nv_window_cfg_t{
        .title = "Todo App",
        .width = 960,
        .height = 720,
        .min_width = 0,
        .min_height = 0,
        .resizable = 1,
        .frameless = 0,
        .transparent = 0,
        .devtools = 1,
        .modal = 0,
    };

    const win = nv.nv_window_create(app, &cfg) orelse {
        std.debug.print("[todo_app] nv_window_create failed\n", .{});
        return;
    };

    g_ctx = .{ .store = store };

    nv.nv_on_ready(win, onReady, null);
    nv.nv_on_message(win, onMessage, &g_ctx);
    nv.nv_window_on_close(win, onClose, @ptrCast(app));

    nv.nv_load_html_ref(
        win,
        @ptrCast(&embed.nv_todo_html_data[0]),
        embed.nv_todo_html_len,
        "about:blank",
    );
    nv.nv_window_show(win);

    nv.nv_app_run(app);
}

File: examples/todo_app/c_todo/src/todo_app.c
/*
 * todo_app.c — Vue UI in embedded HTML + SQLite via todo_bridge IPC.
 */

#include "todo_app.h"
#include "todo_bridge.h"
#include "todo_html.h"
#include "todo_store.h"

#include "nv.h"
#include "nv_arena.h"

#include <stdio.h>
#include <string.h>

typedef struct {
  nv_app_t*      app;
  nv_window_t*   window;
  todo_store_t*  store;
  nv_arena_t*    bridge_arena;
} app_ctx_t;

static void emit_to_js(void* userdata, const char* event, const char* json) {
  nv_window_t* w = (nv_window_t*)userdata;
  if (!w || !event || !json) return;
  nv_send(w, event, json);
}

static void on_message(nv_window_t* window, const char* event, const char* json, void* userdata) {
  app_ctx_t* ctx = (app_ctx_t*)userdata;
  if (!ctx || !ctx->store || !window || !json || !ctx->bridge_arena) return;
  if (!event || strcmp(event, "todo") != 0) return;
  nv_arena_reset(ctx->bridge_arena);
  todo_bridge_handle_wire(ctx->store, ctx->bridge_arena, json, emit_to_js, window);
}

static void on_ready(nv_window_t* window, void* userdata) {
  (void)window;
  (void)userdata;
  printf("[todo_app] webview ready\n");
  fflush(stdout);
}

static void on_close(nv_window_t* window, void* userdata) {
  (void)window;
  nv_app_quit((nv_app_t*)userdata);
}

int todo_app_main(int argc, char** argv) {
  const char* db_path = (argc >= 2 && argv[1] && argv[1][0]) ? argv[1] : "./todo_app.db";

  todo_store_t* store = todo_store_open(db_path);
  if (!store) {
    fprintf(stderr, "[todo_app] todo_store_open failed for %s\n", db_path);
    return 1;
  }

  nv_app_t* app = nv_app_create();
  if (!app) {
    fprintf(stderr, "[todo_app] nv_app_create failed\n");
    todo_store_close(store);
    return 1;
  }

  nv_window_cfg_t wcfg;
  memset(&wcfg, 0, sizeof(wcfg));
  wcfg.title     = "Todo App";
  wcfg.width     = 960;
  wcfg.height    = 720;
  wcfg.resizable = 1;
  wcfg.devtools  = 1;

  nv_window_t* win = nv_window_create(app, &wcfg);
  if (!win) {
    fprintf(stderr, "[todo_app] nv_window_create failed\n");
    todo_store_close(store);
    nv_app_destroy(app);
    return 1;
  }

  nv_arena_t* br = nv_arena_create(256 * 1024);
  if (!br) {
    fprintf(stderr, "[todo_app] bridge arena failed\n");
    nv_window_destroy(win);
    todo_store_close(store);
    nv_app_destroy(app);
    return 1;
  }

  app_ctx_t ctx = { .app = app, .window = win, .store = store, .bridge_arena = br };

  nv_on_ready(win, on_ready, &ctx);
  nv_on_message(win, on_message, &ctx);
  nv_window_on_close(win, on_close, app);
  nv_load_html_ref(win, nv_todo_html(), nv_todo_html_len(), "about:blank");
  nv_window_show(win);

  nv_app_run(app);

  nv_arena_destroy(br);
  todo_store_close(store);
  nv_app_destroy(app);
  return 0;
}
